tfhe/src/integer/server_key/radix_parallel/cmux.rs

```diff
@@ -0,0 +1,232 @@
+    where
+        F: Fn(u64) -> bool + Send + Sync + Copy,
+    {
+        let inverted_predicate = |x| !predicate(x);
+
+        // Although our mul algorithm has special path for when rhs or lhs is a boolean value,
+        // we don't call it as for the ct_false we would need an extra pbs to 'invert' the
+        // ciphertext from true to false.
+        let (mut true_ct, false_ct) = rayon::join(
+            move || {
+                let mut true_ct = true_ct.clone();
+                self.zero_out_if(&mut true_ct, condition_block, inverted_predicate);
+                true_ct
+            },
+            move || {
+                let mut false_ct = false_ct.clone();
+                self.zero_out_if(&mut false_ct, condition_block, predicate);
+                false_ct
+            },
+        );
+        // If the condition was true, true_ct will have kept its value and false_ct will be 0
+        // If the condition was false, true_ct will be 0 and false_ct will have kept its value
+        true_ct
+            .blocks
+            .par_iter_mut()
+            .zip(false_ct.blocks.par_iter())
+            .for_each(|(lhs_block, rhs_block)| {
+                self.key.unchecked_add_assign(lhs_block, rhs_block);
+                self.key.message_extract_assign(lhs_block)
+            });
+        true_ct
+    }
+
+    /// This function takes a ciphertext encrypting any integer value
+    /// and block encrypting a boolean value (0 or 1).
+    ///
+    /// The input integer ciphertext will have all its block zeroed if condition_block
+    /// encrypts 0, otherwise each block keeps its value.
+    pub(crate) fn zero_out_if_condition_is_false(
+        &self,
+        ct: &mut RadixCiphertext,
+        condition_block: &crate::shortint::Ciphertext,
+    ) {
+        assert_eq!(condition_block.degree, Degree(1));
+
+        self.zero_out_if_condition_equals(ct, condition_block, 0)
+    }
+
+    pub(crate) fn zero_out_if_condition_equals(
+        &self,
+        ct: &mut RadixCiphertext,
+        condition_block: &crate::shortint::Ciphertext,
+        value: u64,
+    ) {
+        assert!(condition_block.degree.0 < condition_block.message_modulus.0);
+        assert!(value < condition_block.message_modulus.0 as u64);
+
+        self.zero_out_if(ct, condition_block, |x| x == value);
+    }
+
+    pub(crate) fn zero_out_if<F>(
+        &self,
+        ct: &mut RadixCiphertext,
+        condition_block: &crate::shortint::Ciphertext,
+        predicate: F,
+    ) where
+        F: Fn(u64) -> bool,
+    {
+        assert!(condition_block.degree.0 < condition_block.message_modulus.0);
+
+        let lut =
+            self.key.generate_lookup_table_bivariate(
+                |block, condition| if predicate(condition) { 0 } else { block },
+            );
+
+        ct.blocks.par_iter_mut().for_each(|block| {
+            self.key
+                .unchecked_apply_lookup_table_bivariate_assign(block, condition_block, &lut);
```

Comment:
if you could take a look at smart_evaluate_bivariate_function_assign

I think the smart version has a potential issue with the factor computation the unchecked version looks fine

Reply:
one or the other there is a degree that looks wrong if the ciphertext has not been cleared

Reply:
the degree is used to generate the accumulator and then comes the code to clean the ciphertexts so I'm guessing something can go wrong

Reply:
yeah it looked odd, I changed it